src/main/cpp/instruments/baseinstrument.cpp

```diff
@@ -77,31 +86,41 @@ void BaseInstrument::updateEvents()
     // or to update event properties responding to tempo changes
     // override this function in your derived class for custom implementations
 
-    if ( _oldTempo != AudioEngine::tempo ) {
-
-        //std::lock_guard<std::mutex> guard( _lock );
-        toggleReadLock( true );
+    if ( _oldTempo == AudioEngine::tempo ) {
+        return;
+    }
 
-        // when tempo has updated, we update the offsets of all associated events
+    //std::lock_guard<std::mutex> guard( _lock );
+    toggleReadLock( true );
 
-        float ratio = _oldTempo / AudioEngine::tempo;
+    // when tempo has updated, we update the offsets of all associated events
+    // note the measure cache remains untouched (nothing changes with regards to
+    // measure separation)
 
-        for ( int i = 0, l = _audioEvents->size(); i < l; ++i )
-        {
-            BaseAudioEvent* event = _audioEvents->at( i );
+    float ratio = _oldTempo / AudioEngine::tempo;
+    _oldTempo   = AudioEngine::tempo;
 
-            auto orgStart  = ( float ) event->getEventStart();
-            auto orgEnd    = ( float ) event->getEventEnd();
-            auto orgLength = ( float ) event->getEventLength();
+    // TODO: there is an issue that creeps in with regards to maintaining an accurate measure
+    // cache, get(Start|End)MeasureForEvent relies on AudioEngine::samples_per_bar so the removal is
+    // calculating for the new start/end measure range, therefor possibly missing the old (prior to
+    // tempo change) range values. Additionally, we might risk undefined behaviour on the read locks
+    // for an already locked mutex. We set a flog to prevent event add/remove changes (triggered by their
+    // repositioning) and invoke a manual flush and recache after all sequenced events have been repositioned.
 
-            event->setEventStart ( orgStart  * ratio );
-            event->setEventLength( orgLength * ratio );
-            event->setEventEnd   (( orgEnd + 1 ) * ratio ); // add 1 to correct for rounding of float
-        }
-        _oldTempo = AudioEngine::tempo;
+    _freezeEvents = true;
 
-        toggleReadLock( false );
+    size_t i = 0, total = _audioEvents->size();
+    for ( ; i < total; ++i ) {
+        _audioEvents->at( i )->repositionToTempoChange( ratio );
     }
```

Comment:
Now only _BaseInstrument_ has its own definition for resynchronizing events after a tempo change. The event type can determine its own implementation (in _repositionToTempoChange()_). Further more, default behaviour has been taken from _SampledInstrument_ : when adjusting an events start position (to compensate for offset shift by tempo change), only the end position is adjusted, leaving the total duration the same (as we want event buffers to complete playback in their entirety).

Only _BaseSynthEvent_ deviates from this norm as its duration is implied by the Sequencer tempo, also adjusting the event length in the proces.